File: Programmers/Level2/올바른괄호.cpp
// 괄호 문제 !!!
// Stack을 활용하는 문제!
#include <string>
#include <iostream>
#include <stack>

using namespace std;

bool solution(string s)
{
    bool answer = true;

    stack<char> words;


    for(int i = 0 ; i < s.length() ; i++){
      if(s[i] == '('){
        words.push('(');
      }else if(s[i] == ')'){
        if(words.empty()){
          answer = false;
          break;
        }else{
          if(words.top() == '('){
            words.pop();
          }
        }
      }
    }

    if(!words.empty()){
      answer = false;
    }

    return answer;
}

File: JavaScript/[1]Programmers/코딩테스트 고득점 Kit/올바른괄호.js
function solution(s){

    let stack = [];

    for(let i = 0 ; i < s.length ; i++) {
        if (s[i] === "(") {
            stack.push("(");
        } else {
            if (stack.length === 0) {
                return false;
            } else {
                stack.pop();
            }
        }
    }
    return stack.length === 0;
}

const s1 = "()()";
const s2 = "(())()"
const s3 = ")()(";
const s4 = "(()(";

console.log(solution(s1));